### PR TITLE
{TESTS RUNNING} Fix Kafka Importer occasionally losing small numbers of rows upon pause

### PR DESCRIPTION
**Pull request by brians127**
This fixes a race between the VoltDB procedure callback threads trying to set the paused offset. It was reliably set to the FIRST value seen, but not reliably set to the LOWEST value seen across all threads.

I introduce a synchronization mechanism to ensure the paused offset is reliable before using it to commit any offsets to Kafka.

Some design decisions behind the PendingWorkTracker object, which should be reusable:
* Routine adjustment of pending callback counters is lockless and fast
* Producer needs to reliably know when all callbacks are done without periodic sleeps
* Need to be able to get the number of callbacks called at the end of the run, regardless of whether or not waitForWorkToFinish() has been called.

### CODE REVIEW

File: src/frontend/org/voltdb/importclient/kafka/KafkaTopicPartitionImporter.java [hunk 1]
```diff
@@ -465,14 +466,18 @@ protected void accept() {
...
-                                messageAndOffset.nextOffset(), cbcnt, m_gapTracker, m_dead, m_pauseOffset);
-                         if (!noTransaction && !callProcedure(invocation, cb)) {
-                              if (isDebugEnabled()) {
-                                 debug(null, "Failed to process Invocation possibly bad data: " + Arrays.toString(params));
-                              }
-                              m_gapTracker.commit(messageAndOffset.nextOffset());
+                                messageAndOffset.nextOffset(), callbackTracker, m_gapTracker, m_dead, m_pauseOffset);
+                        if (!noTransaction) {
+                            if (callProcedure(invocation, cb)) {
+                                callbackTracker.produceWork();
```
Comment (brians127) [inline src/frontend/org/voltdb/importclient/kafka/KafkaTopicPartitionImporter.java]: This line is new. The rest of this change is minor - changing cbcnt to callbackTracker and fixing indentation.

File: src/frontend/org/voltdb/importclient/kafka/KafkaTopicPartitionImporter.java [hunk 2]
```diff
@@ -494,14 +499,21 @@ protected void accept() {
...
                 }
                 if (shouldCommit()) {
-                    commitOffset();
+                    long pausedOffset = -1; // ignores paused offset since we don't want to wait for all callbacks to complete
+                    commitOffset(pausedOffset);
                 }
             }
         } catch (Exception ex) {
             error(ex, "Failed to start topic partition fetcher for " + m_topicAndPartition);
+            ex.printStackTrace();
```
Comment (brians127) [inline src/frontend/org/voltdb/importclient/kafka/KafkaTopicPartitionImporter.java]: I put this in for debugging and never hit it - suggestions? Remove change or dump it to the log?

File: src/frontend/org/voltdb/importclient/kafka/KafkaTopicPartitionImporter.java [hunk 3]
```diff
@@ -494,14 +499,21 @@ protected void accept() {
                         }
                 }
                 if (shouldCommit()) {
-                    commitOffset();
+                    long pausedOffset = -1; // ignores paused offset since we don't want to wait for all callbacks to complete
```
Comment (akhanzode) [inline src/frontend/org/voltdb/importclient/kafka/KafkaTopicPartitionImporter.java]: turn this into boolean and pass true.

File: src/frontend/org/voltdb/importclient/kafka/KafkaTopicPartitionImporter.java [hunk 4]
```diff
@@ -494,14 +499,21 @@ protected void accept() {
...
                 }
             }
         } catch (Exception ex) {
             error(ex, "Failed to start topic partition fetcher for " + m_topicAndPartition);
+            ex.printStackTrace();
         } finally {
-            //Dont care about return as it wil force a commit.
-            commitOffset();
+            long pausedOffset = m_pauseOffset.get();
+            if (pausedOffset != -1) {
```
Comment (akhanzode) [inline src/frontend/org/voltdb/importclient/kafka/KafkaTopicPartitionImporter.java]: Put comment here.

**Comment by brians127** [comment]
JIRA ticket is incorrect - should be ENG-12348. I'm closing this pull request and opening a new one with the same code but the correct ticket. Changes as per code review feedback will be discussed and addressed there.

**Pull request closed without merging**